daemon/cluster/listen_addr.go

```diff
@@ -0,0 +1,163 @@
+package cluster
+
+import (
+	"errors"
+	"fmt"
+	"net"
+)
+
+func resolveListenAddr(specifiedAddr string, defaultListenAddr string) (string, error) {
+	specifiedHost, specifiedPort, err := net.SplitHostPort(specifiedAddr)
+	if err != nil {
+		return "", fmt.Errorf("could not parse listen address %s", specifiedAddr)
```

Comment:
@icecrime: The issue is that the port is chosen client-side. So currently `0.0.0.0:2377` is a valid value for `ListenAddr`, but an empty string is not.

We could potentially move the default port selection to the daemon side, but this has UX impacts. From the client side, the default of `--listen-addr` would change to an empty value, so it would not be as obvious what port number gets used if you don't specify anything. We would probably want to make `--swarm-default-listen-addr` take an optional port value, and then we would have to handle corner cases like no port number being specified in either argument.

Ultimately I think the choice of port belongs on the client side, but I'm curious what you think.